examples/all-clusters-app/all-clusters-common/src/WhmDelegateImpl.cpp

```diff
@@ -0,0 +1,541 @@
+
+    ChipLogProgress(AppServer, "HandleCancelBoost");
+
+    if (mBoostState == BoostStateEnum::kActive)
+    {
+        SetBoostState(BoostStateEnum::kInactive);
+
+        DeviceLayer::SystemLayer().CancelTimer(BoostTimerExpiry, this);
+
+        if (mpWhmManufacturer != nullptr)
+        {
+            status = mpWhmManufacturer->BoostCommandCancelled();
+        }
+        else
+        {
+            status = Status::InvalidInState;
+            ChipLogError(AppServer, "HandleCancelBoost: mpWhmManufacturer == nullptr");
+        }
+
+        status = CheckIfHeatNeedsToBeTurnedOnOrOff();
+    }
+
+    return status;
+}
+
+/*********************************************************************************
+ *
+ * Methods implementing the ModeBase::Delegate interface
+ *
+ *********************************************************************************/
+
+CHIP_ERROR WaterHeaterManagementDelegate::Init()
+{
+    return CHIP_NO_ERROR;
+}
+
+void WaterHeaterManagementDelegate::HandleChangeToMode(uint8_t NewMode, ModeBase::Commands::ChangeToModeResponse::Type & response)
+{
+    response.status = to_underlying(ModeBase::StatusCode::kSuccess);
+}
+
+CHIP_ERROR WaterHeaterManagementDelegate::GetModeLabelByIndex(uint8_t modeIndex, chip::MutableCharSpan & label)
+{
+    if (modeIndex >= ArraySize(kModeOptions))
+    {
+        return CHIP_ERROR_PROVIDER_LIST_EXHAUSTED;
+    }
+
+    return chip::CopyCharSpanToMutableCharSpan(kModeOptions[modeIndex].label, label);
+}
+
+CHIP_ERROR WaterHeaterManagementDelegate::GetModeValueByIndex(uint8_t modeIndex, uint8_t & value)
+{
+    if (modeIndex >= ArraySize(kModeOptions))
+    {
+        return CHIP_ERROR_PROVIDER_LIST_EXHAUSTED;
+    }
+
+    value = kModeOptions[modeIndex].mode;
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR WaterHeaterManagementDelegate::GetModeTagsByIndex(uint8_t modeIndex, DataModel::List<ModeTagStructType> & tags)
+{
+    if (modeIndex >= ArraySize(kModeOptions))
+    {
+        return CHIP_ERROR_PROVIDER_LIST_EXHAUSTED;
+    }
+
+    if (tags.size() < kModeOptions[modeIndex].modeTags.size())
+    {
+        return CHIP_ERROR_INVALID_ARGUMENT;
+    }
+
+    std::copy(kModeOptions[modeIndex].modeTags.begin(), kModeOptions[modeIndex].modeTags.end(), tags.begin());
+    tags.reduce_size(kModeOptions[modeIndex].modeTags.size());
+
+    return CHIP_NO_ERROR;
+}
+
+/*********************************************************************************
+ *
+ * WaterHeaterManagementDelegate specific methods
+ *
+ *********************************************************************************/
+
+void WaterHeaterManagementDelegate::SetWaterTemperature(uint16_t waterTemperature)
+{
+    // This method assumed 100% of the water in the tank has reached the waterTemperature specified
+    mHotWaterTemperature = waterTemperature;
+
+    // Do not change mTankPercentage if the kTankPercent feature is not supported
+    if (mpWhmInstance != nullptr && mpWhmInstance->HasFeature(Feature::kTankPercent))
```

Comment:
SetWaterTemperature sets 100% of the tank to the temperature specified. Therefore mTankPercentage needs to be 100%

Reply:
I think the logic here should calculate the tank percentage as follows (for a single probe):

 mTankPercentage = ((mHotWaterTemperature - 2000) / (mTargetTemperature - 2000)) * 100; // where 2000 = 20'C (the assumed cold water temperature)
 mTankPercentage = min(100, mTankPercentage);
 mTankPercentage = max(0, mTankPercentage); 


i.e. the tankPercentage is just another way of expressing how hot the water is (and it assumes that a single temperature probe is measuring the average water temperature) which in reality the water isn't mixed and you get stratification layers where it suddenly changes from hot to cold. 

More sophisticated tanks have multiple probes mounted vertically and it can check where the layer of hot and cold water is.

Reply:
So I interpret the member variables:
**`WaterHeaterManagementDelegate::mWaterTemperature `
`WaterHeaterManagementDelegate::::mTankPercentage`**

as

mTankPercentage of the total tank volume is at the temperature mWaterTemperature